README (#42)
thefuntasty/CellKit: README.md

![Cocoapods platforms](https://img.shields.io/cocoapods/p/CellKit.svg)
![License](https://img.shields.io/cocoapods/l/CellKit.svg)
![Continuous integration](https://img.shields.io/bitrise/57c06040e101a852.svg?token=xvYC9NW6ZszIzFdjkapmVg)

> CellKit is a swift package that simplifies the workflow with cells in collection views like UITableViews and UICollectionView with MVVM architecture. No more registering cell classes or xibs, no more dequing cells and setting its view models. CellKit handles this all.

Reviewer: ```suggestion
CellKit is a Swift package that simplifies the workflow with cells in UITableView and UICollectionView. No more registering cell classes or XIBs, no more dequeueing cells and setting its view models. CellKit handles this all.
```

## Installation
### Swift Package
```
.package(url: "https://github.com/futuredapp/CellKit", from: "0.8.0")
```
### CocoaPods
```

> pod 'CellKit', '~> 0.8.0'

Reviewer: ```suggestion
pod 'CellKit', '~> 0.8'
```

```

## Features

> CellKit provides a dataSource and a section model which you fill with your cells, headers and footers. All you're left to do is to define your cell View and your cell ViewModel with CellKit protocols and CellKit will handle the rest.

Reviewer: ```suggestion
CellKit provides a data source and a section model which you fill with your cells, headers and footer models. All you're left to do is to define your cell view and your cell model with protocol fonformance to CellConvertible and CellConfigurable and CellKit will handle the rest.
```

> ### DataSource

Reviewer: ```suggestion
### Data source
```

DataSource and section model are pretty straight forward. CellKit comes with `CellModelDataSource` and `GenericCellModelSection` which represents your UITableView/UICollectionView cell structure. 

#### CellModelDataSource
`CellModelDataSource` is an open class which means that you can make a subclass a override its behaviour to suit your needs.
`CellModelDataSource` only accepts an array of `GenericCellModelSection` objects.

#### GenericCellModelSection
`GenericCellModelSection` represents a section in tableView/collectionView. It accepts following arguments:

> 

Reviewer: Missing arguments.

### Cell protocols
In order for your cells and cell view models to work with CellKit, they have to conform to three protocols: `CellConfigurable`, `CellConvertible` and `CellModel`

#### CellConfigurable protocol

> Provides a `configuration(model:)` method which gets call when a tableview dequest a new reusable cell. This protocol is for cell views only.

Reviewer: ```suggestion
Provides a `configure(model:)` method which gets call when a UITableView requests a new reusable cell. This protocol is for cell views only.
```

> 

Reviewer: Missing arguments.

    
    // MARK: CellConvertible
    typealias Cell = PrototypeCell
    var usesNib: Bool { false }
    var registersLazily: Bool { false }
}
```
### With cell defined in code
cell:
```swift
class CodeCell: UITableViewCell, CellConfigurable {
    var label: UILabel = UILabel()
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        addSubview(label)
        label.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            topAnchor.constraint(equalTo: label.topAnchor, constant: -16),
            bottomAnchor.constraint(equalTo: label.bottomAnchor, constant: 16),
            leftAnchor.constraint(equalTo: label.leftAnchor, constant: -16),
            heightAnchor.constraint(equalToConstant: 64)
        ])
    }
    
    func configure(with model: CodeCellViewModel) {
        label.text = "\(model.name)"
    }
}
```
model:
```swift
struct CodeCellViewModel: CellConvertible, CellModel {
    let name: String
    
    // MARK: CellConvertible
    typealias Cell = CodeCell
    var usesNib: Bool { false }
    var cellHeight: Double = 64
}

> ```

Reviewer: Please add contributing guide, maintainers and license.

![Cocoapods platforms](https://img.shields.io/cocoapods/p/CellKit.svg)
![License](https://img.shields.io/cocoapods/l/CellKit.svg)
![Continuous integration](https://img.shields.io/bitrise/57c06040e101a852.svg?token=xvYC9NW6ZszIzFdjkapmVg)

CellKit is a Swift package that streamlines the workflow with cells in UITableView and UICollectionView. No more registering cell classes or XIBs, no more dequeueing cells and setting its view models. CellKit handles this all.

## Installation
### Swift Package
Add following line to your swift package dependencies, or in Xcode, go to `File -> Swift Packages -> Add Package Dependency` and type in URL address of this repository.
```
.package(url: "https://github.com/futuredapp/CellKit", from: "0.8.0")
```
### CocoaPods

> Add following line to your  `Podfile` and then run `bundle exec pod install`

Reviewer: I wouldn't force bundler usage to all the users.

```suggestion
Add following line to your `Podfile` and then run `pod install`
```

```
pod 'CellKit', '~> 0.8'
```

## Usage
CellKit provides a data source and a section model which you fill with your cells, headers and footer models. All you're left to do is to define your cell view and your cell model with protocol conformance to CellConvertible and CellConfigurable and CellKit will handle the rest.

### 1. step: Set a DataSource

> CellKit provides a `CellModelDataSource` datasource and `GenericCellModelSection` section, which define your `UITableView`/`UICollectionView` cell structure. You can always subclass `CellModelDataSource` and override it's methods to suit your needs.  

Reviewer: I think it's cleat without the repetition.

```suggestion
CellKit provides a `CellModelDataSource` and `GenericCellModelSection`, which define your `UITableView`/`UICollectionView` cell structure. You can always subclass `CellModelDataSource` and override its methods to suit your needs.  
```

> Here's an example of typical CellKit datasource usage:

Reviewer: ```suggestion
Here's an example of typical CellKit data source usage:
```

```swift

> lazy var dataSource: CellModelDataSource = {
>     CellModelDataSource([
>         GenericCellModelSection(arrayLiteral:
>             PrototypeCellViewModel(name: "Prototype")
>         ),
>         GenericCellModelSection(arrayLiteral:
>             CodeCellViewModel(name: "Code")
>         ),
>         GenericCellModelSection(arrayLiteral:
>             XIBCellViewModel(name: "XIB")
>         )
>     ])
> }()

Reviewer: I wouldn't confuse the users with the lazy definition. It doesn't make much sense in this particular implementation.

```suggestion
let dataSource = CellModelDataSource([
    GenericCellModelSection(arrayLiteral:
        PrototypeCellViewModel(name: "Prototype")
    ),
    GenericCellModelSection(arrayLiteral:
        CodeCellViewModel(name: "Code")
    ),
    GenericCellModelSection(arrayLiteral:
        XIBCellViewModel(name: "XIB")
    )
])
```

tableView.dataSource = dataSource
collectionView.dataSource = dataSource
```

### 2. step: Create your cell and CellModel
CellKit support wide variety of cell declarations including `.xib` files , `UITableView` interface builder prototype cells and Cells written entirely in code.  
*Please note that your `.xib` file, Cell subclass and Cell identifier have to to have the same name.* It is possible to not use the same identifier, but it is not recommended.  
Pro tip: You can use our [custom teplates](https://github.com/futuredapp/MVVM-C-Templates) to generate Cell with CellKit protocols in just a few clicks.

> ### 3. Implement CellKit protocols

Reviewer: ```suggestion
### 3. step: Conform to CellKit protocols
```

In order for your cells and cell view models to work with CellKit, they have to conform to these protocols:

#### CellConfigurable
This is a protocol for your *Cell*. This protocol provides a `configure(model:)` method which gets call when a tableview request a reusable cell and is used to distribute your model to your cells. 
```swift
class XIBCell: UITableViewCell, CellConfigurable {
    @IBOutlet private weak var label: UILabel!
}

extension XIBCell: CellConfigurable {
    func configure(with model: XIBCellViewModel) {
        label.text = model.name
    }
}
```

#### CellModel

> Protocol for your cell model. Use this procotol to specify your cell configuration such as height xib location, whether the cell can be highlighted, etc.. 

Reviewer: ```suggestion
Protocol for your cell model. Use this procotol to specify your cell configuration such as height xib location, whether the cell can be highlighted, etc.
```

| nib               | UINib?   | xib with `cellClass` name. Or nil if `usesNib` is false. | An `UINib` reference of your .xib file containing you view                    |
| cellClass         | AnyClass |                                                          | A class reference to views class.                                           |
| reuseIdentifier   | String   |                                                          | a unique re-use identifier                                                  |
| cellHeight        | Double   | 44.0                                                     | hight for cell                                                              |
| highlighting      | Bool     | true                                                     | Indicates whether cell can be highlighted                                   |
| separatorIsHidden | Bool     | false                                                    | Indicates whether should hide separator                                     |

#### CellConvertible 

> This protocol extends CellModel with associated type and thus can provide a default `cellClass`  and  `reuseIdentifier` value based on type's name.    

Reviewer: ```suggestion
This protocol extends CellModel with associated type and thus can provide a default `cellClass` and  `reuseIdentifier` value based on type's name.    
```

It's espeatially handy when you declare your cell in XIB, becouse all you need to do is define its associated type and CellKit will provide the rest. Take a look at `XIB cell` example.

## Examples
### XIB cell
cell:
```swift
class XIBCell: UITableViewCell, CellConfigurable {
    @IBOutlet private weak var label: UILabel!
    func configure(with model: XIBCellViewModel) {
        label.text = "\(model.name)"
    }
}
```
model:
```swift
struct XIBCellViewModel: CellConvertible, CellModel {
    let name: String
    
    // MARK: CellConvertible
    typealias Cell = XIBCell
}
```

### Storyboard prototype cell
cell:
```swift
class PrototypeCell: UITableViewCell, CellConfigurable {
    @IBOutlet private weak var label: UILabel!
    
    func configure(with model: PrototypeCellViewModel) {
        label.text = "\(model.name)"
    }
}
```
model:
```swift
struct PrototypeCellViewModel: CellConvertible, CellModel {
    let name: String
    
    // MARK: CellConvertible
    typealias Cell = PrototypeCell
    var usesNib: Bool { false }
    var registersLazily: Bool { false }
}
```
### Cell defined in code
cell:
```swift
class CodeCell: UITableViewCell, CellConfigurable {

>     var label: UILabel = UILabel()

Reviewer: ```suggestion
    let label: UILabel = UILabel()
```

    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        addSubview(label)
        label.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            topAnchor.constraint(equalTo: label.topAnchor, constant: -16),
            bottomAnchor.constraint(equalTo: label.bottomAnchor, constant: 16),
            leftAnchor.constraint(equalTo: label.leftAnchor, constant: -16),
            heightAnchor.constraint(equalToConstant: 64)
        ])
    }
    
    func configure(with model: CodeCellViewModel) {
        label.text = "\(model.name)"
    }
}
```
model:
```swift
struct CodeCellViewModel: CellConvertible, CellModel {
    let name: String
    
    // MARK: CellConvertible
    typealias Cell = CodeCell

>     var usesNib: Bool { false }
>     var cellHeight: Double = 64

Reviewer: Let's simplify it a bit.

```suggestion
    let usesNib = false
    let cellHeight = 64.0
```
